src/utils/daos_hdlr.c

```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		 */
+		rc = daos_cont_open(ca.dst_poh, ca.dst_c_uuid, DAOS_COO_RW,
+				    &ca.dst_coh, &dst_cont_info, NULL);
+		if (rc == 0) {
+			fprintf(stderr, "This destination container already "
+				"exists, please provide a destination "
+				"container uuid that does not exist already, "
+				"or provide an existing pool or new UNS path "
+				"of the form:\n\t--dst </$pool> | "
+				"<path/to/uns>\n");
+			/* disconnect from only destination and exit */
+			rc = daos_cont_close(ca.dst_coh, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to close destination "
+					"container (%d)\n", rc);
+				D_GOTO(out, rc);
+			}
+			rc = daos_pool_disconnect(ca.dst_poh, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to disconnect from "
+					"destination pool "DF_UUIDF ": %s "
+					"(%d)\n", DP_UUID(ca.dst_p_uuid),
+					d_errdesc(rc), rc);
+				D_GOTO(out, rc);
+			}
+			D_GOTO(out, rc = 1);
+		}
+	}
+
+	rc = dm_connect(is_posix_copy, &dst_cp_type, &src_cp_type,
+			&ca, ap->sysname, ap->dst, &src_cont_info,
+			&dst_cont_info);
+	if (rc != 0) {
+		D_GOTO(out_disconnect, rc);
+	}
+	rc = daos_cont_create_snap_opt(ca.src_coh,
+				       &epoch, NULL,
+			       DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT, NULL);
+	if (rc) {
+		fprintf(stderr, "failed to create snapshot: %d\n", rc);
+		D_GOTO(out_disconnect, rc);
+	}
+	rc = daos_oit_open(ca.src_coh, epoch, &toh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to open object iterator\n");
+		D_GOTO(out_snap, rc);
+	}
+	memset(&anchor, 0, sizeof(anchor));
+	while (!daos_anchor_is_eof(&anchor)) {
+		oids_nr = OID_ARR_SIZE;
+		rc = daos_oit_list(toh, oids, &oids_nr, &anchor, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to list objects\n");
+			D_GOTO(out_oit, rc);
+		}
+
+		/* list object ID's */
+		for (i = 0; i < oids_nr; i++) {
+			rc = daos_obj_open(ca.src_coh, oids[i],
+					   0, &oh, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to open source "
+					"object\n");
+				D_GOTO(out_oit, rc);
+			}
+			rc = daos_obj_open(ca.dst_coh, oids[i], 0,
+					   &dst_oh, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to open destination "
+					"object\n");
+				D_GOTO(err_dst, rc);
+			}
+			rc = cont_clone_list_dkeys(&oh, &dst_oh);
+			if (rc != 0) {
+				fprintf(stderr, "failed to list keys\n");
+				D_GOTO(err_obj, rc);
+			}
+			rc = daos_obj_close(oh, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to close source "
+					"object: %d\n", rc);
+				D_GOTO(out_oit, rc);
+			}
+			rc = daos_obj_close(dst_oh, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to close destination "
+				"object: %d\n", rc);
+				D_GOTO(err_dst, rc);
+			}
+		}
+	}
+	D_GOTO(out_oit, rc);
+err_obj:
+	rc = daos_obj_close(oh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to close source object: %d\n", rc);
+	}
+	rc = daos_obj_close(dst_oh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to close destination object: %d\n", rc);
+	}
+err_dst:
+	rc = daos_obj_close(oh, NULL);
```

Comment:
@dsikich Maybe it's okay, but if we goto err_obj, then both err-obj and err_dst try to close `oh`

Reply:
@daltonbohning I don't know if it is not ok, but not really sure a way around that for that case.

Reply:
@dsikich Could we just remove it from err_obj and let it fall through to err_dst?

Reply:
@daltonbohning yeah I guess that would close them both in if both objects are already opened